File: External-Tutorial-SubComponentBuilder/src/main/java/zorail/rohan/com/subcomponent/SecondActivity.java
package zorail.rohan.com.subcomponent;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import javax.inject.Inject;

import zorail.rohan.com.subcomponent.component.NetComponent;
import zorail.rohan.com.subcomponent.module.NetModule;

public class SecondActivity extends AppCompatActivity {

    @Inject
    SharedPreferences preferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_second);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        NetComponent.Builder builder = (NetComponent.Builder)((MyApp)getApplication()).getComponent().subcomponentBuilders().get(NetComponent.Builder.class).get();
        builder.netModule(new NetModule()).build().inject(this);

        Toast.makeText(this,preferences.getString("Name",""),Toast.LENGTH_LONG).show();

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
    }

}

File: Tutorial6-3SubComponentFactory/src/main/java/com/smarttoolfactory/tutorial6_3subcomponentfactory/SecondActivity.kt
package com.smarttoolfactory.tutorial6_3subcomponentfactory

import android.content.SharedPreferences
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.smarttoolfactory.tutorial6_3subcomponentfactory.model.MySharedPreferences
import com.smarttoolfactory.tutorial6_3subcomponentfactory.model.SensorController
import com.smarttoolfactory.tutorial6_3subcomponentfactory.model.SingletonObject
import com.smarttoolfactory.tutorial6_3subcomponentfactory.model.ToastMaker
import javax.inject.Inject

/**
 * ToastMakerModule and MySharedPreferencesModule are scoped with @FragmentScope
 */
class SecondActivity : AppCompatActivity() {

    // Injected from ApplicationModule
    @Inject
    lateinit var sharedPreferences: SharedPreferences

    // Injected from ToastMakerModule @FragmentScope
    @Inject
    lateinit var toastMaker: ToastMaker

    // Injected from MySharedPreferencesModule @FragmentScope
    @Inject
    lateinit var mySharedPreferences: MySharedPreferences

    // Injected via Constructor Injection NOT Singleton
    @Inject
    lateinit var sensorController: SensorController

    // 🔥 Injected via Constructor Injection with @Singleton scope
    @Inject
    lateinit var singletonObject: SingletonObject

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_second)

        initInjection()

        findViewById<TextView>(R.id.tvInfo).text =
                "ApplicationModule sharedPreferences: ${sharedPreferences.hashCode()}\n" +
                        "@FragmentScope MySharedPreferences: ${mySharedPreferences.hashCode()}\n" +
                        "@FragmentScope ToastMaker: ${toastMaker.hashCode()}\n" +
                        "Constructor Un-scoped sensorController: ${sensorController.hashCode()}\n"+
                        "Constructor @Singleton singletonObject: ${singletonObject.hashCode()}"

        toastMaker.showToast("SharedPreferences: $sharedPreferences")

    }

    private fun initInjection() {

        val appComponent = (application as MyApplication).appComponent


        // 🔥SubComponent Builder
//        val toastMakerSubComponent = appComponent
//                .toastMakerBuilder()
//                .context(this)
//                .build()

        // 🔥 SubComponent Factory
        val toastMakerSubComponent = appComponent
                .toastAndSharedPreferencesFactory
                .create(this)

        toastMakerSubComponent.inject(this)


    }
}